pywbem_mock/_namespaceprovider.py

```diff
@@ -0,0 +1,290 @@
+    # these class level attributes must exist to define the provider
+    provider_type = 'instance'
+    provider_classnames = 'CIM_Namespace'
+
+    def __init__(self, cimrepository):
+        """
+        Set up the cim repository
+
+        """
+        super(CIMNamespaceProvider, self).__init__(cimrepository)
+
+        # Set by install_provider since __init__ does not have access to
+        # conn
+        self.interop_namespaces = None
+
+    def __repr__(self):
+        return _format(
+            "CIMNamespaceProvider("
+            "provider_type={s.provider_type}, "
+            "provider_classnames={s.provider_classnames})",
+            s=self)
+
+    def CreateInstance(self, namespace, NewInstance):
+        """
+        Create an instance of the CIM_Namespace class with the required
+        parameters.  If the namespace defined in NewInstance.name does not
+        exist, it creates the namespace in the cimrepository.
+        """
+        self.validate_namespace(namespace)
+
+        new_instance = deepcopy(NewInstance)
+
+        if not self.is_interop_namespace(namespace):
+            raise CIMError(
+                CIM_ERR_INVALID_PARAMETER,
+                _format(
+                    "Instances of {0!A} are only allowed in the interop "
+                    "namespace for the server. {1!A} is not an interop "
+                    "namespace. It must be one of {2!A} ",
+                    NewInstance.classname, namespace,
+                    ", ".join(self.cimrepository.interop_namespace_names())))
+
+        try:
+            new_namespace = new_instance['Name']
+        except KeyError:
+            raise CIMError(
+                CIM_ERR_INVALID_PARAMETER,
+                _format("Namespace creation via CreateInstance: "
+                        "Missing 'Name' property in the {0!A} instance ",
+                        new_instance.classname))
+
+        # Normalize the namespace name
+        # TODO: Somehow we need to get this out of the way.
+        new_namespace = new_namespace.strip('/')
+
+        # Write it back to the instance in case it was changed
+        new_instance['Name'] = new_namespace
+
+        # These values must match those in
+        # tests/unittest/utils/wbemserver_mock.py.
+        try:
+            if new_instance['CreationClassName'] != new_instance.classname \
+                or new_instance['ObjectManagerName'] != OBJECTMANAGERNAME \
+                or new_instance['ObjectManagerCreationClassName'] != \
+                'CIM_ObjectManager' \
+                or new_instance['SystemName'] != SYSTEMNAME \
+                or new_instance['SystemCreationClassName'] != \
+                    SYSTEMCREATIONCLASSNAME:
+
+                raise CIMError(
+                    CIM_ERR_INVALID_PARAMETER,
+                    _format("Namespace creation via CreateInstance: "
+                            "Invalid property for CIM_Instance {0!A}",
+                            new_instance))
```

Comment:
Can we not default these properties, if not set?
If they are set, should we not mostly accept them as set, and only check CreationClassName?

Reply:
TODO

Reply:
Fix per comments.